feat: Kubewarden stack telemetry.

jvanz opened the pull request:
## Description

Adds initial draft of the RFC describing how Kubearden can send environment telemetry to central location for further analisys.


Fix https://github.com/kubewarden/kubewarden-controller/issues/1167

jvanz committed:
feat: Kubewarden stack telemetry.

Adds initial draft of the RFC describing how Kubearden can send
environment telemetry to central location for further analisys.

Signed-off-by: José Guilherme Vanz <jguilhermevanz@suse.com>

flavio commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+| Feature Name | Kubewarden environment telemetry |
+| Start Date   | 2025-08-28                       |
+| Category     | telemetry                        |
+| RFC PR       | [fill this in after opening PR]  |
```
This can be filled now

viccuad replied:
Reopening, we can set the state to IN-REVIEW or so, while we wait for CNCF approval.

flavio commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+
+As a Kubewarden user, I want to ensure that the process of sending telemetry
+data does not consume excessive resources in my environment.
+
```
Maybe we can add a user story about a user who doesn't want telemetry data to be sent at all.

flavio commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+about user applications, but rather to learn more about the typical
+environments where Kubewarden is deployed. This data will help guide the
+project's future development.
+
```
I would add a paragraph stressing out this can be turned off

olblak replied:
It depends if we want to opt-in or opt-out by default.
As a project maintainer I would prefer to opt-in by default, it provide more data
As a project user, I would prefer to opt out by default :)

Ultimately I am fine with both but we need to take a decision on this

flavio commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+
+- Potential for breaking changes in the OpenTelemetry specification or
+  collector components.
+- If a user already has a custom OpenTelemetry collector, integrating our
+  telemetry pipeline might be complex or they might filter our data out.
```
I think we can run as many collectors as we want. Meaning, that we could always deploy our own collector that sends the telemetry data to the remote location, regardless of the user having its own collector.

jvanz replied:
Yes. Considering that we allow users to use their own [custom Otel collector](https://docs.kubewarden.io/howtos/telemetry/custom-otel-collector#install-kubewarden-stack),  this would require the controller to send metrics to two different Otel collector. 

But as Otel integration will not be done, this is not relevant anymore.

flavio reviewed (commented):
Thanks for having written this RFC.

~~I think the best option would be to go with the OpenTelemetry approach.~~

I've changed my mind, I think the best option is to go with the second approach since it would allow us to reuse the same technology Longhorn uses.

olblak commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+> As a CNCF project, it is necessary to get approval from the CNCF to implement either of the following options.
+> [Telemetry Data Collection and Usage Policy | Linux Foundation](https://www.linuxfoundation.org/legal/telemetry-data-policy)
+
+## Option 1: OpenTelemetry Integration
```
My main concern with this approach, is that we expect end users to provide extra effort to provide telemetry, which won't be done by most of end-users. If we don't opt in by default telemetry then we need to be sure that it can be enabled in the most straightforward way.

olblak commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+// to a metric observer or recorder.
+commonAttrs := []attribute.KeyValue{
+    attribute.String("policy_server_id", policy.GetPolicyServer().GetUID()),
+    attribute.String("policy_server_image", policy.GetPolicyServer().GetImageDigest()),
```
Are we planning to store policy names? This could lead to end user data leak depending on the name 
for example:

`oci://registry.corporate.com/mysupersecretpolicy`

olblak replied:
Re-reading my comment, I realize that `policy_server_image` is for the server image but still what if the end user rebuilds the kubewarden images, is that an information we should track here

jvanz replied:
Answering the question, the metric should not contains the policy server name or the policy name. To do not expose the registry url and other local information by accident. This proposal, which was not select by the team, would send only the oci artifact digest. No names. But this is not selected integration option.

olblak commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+processors:
+  # Control how much time we can wait before sending the metric to the remote Otel collector
+  batch:
+    timeout: 1h
```
If we send the metrics every hour then it sounds more like real time monitoring

jvanz replied:
This is a example configuration. It can be bigger. I just use the Longhorn [example](https://github.com/longhorn/upgrade-responder?tab=readme-ov-file#how-it-works). 

> Server will expect each instance of the application makes a request once per hour.

jvanz committed:
fix: discart OpenTelemetry integration option.

During a conversation, the Kubewaredn team decided to not do the
integration with OpenTelemetry and adoppt the integration with the
Longhorn upgrade-responder server. Therefore, this commit moves that
option to the "Alternatives" section.

Signed-off-by: José Guilherme Vanz <jguilhermevanz@suse.com>

jvanz committed:
feat: desired metrics section.

Adds a section describing the metrics that the Kubewarden team should
would like to have.

Signed-off-by: José Guilherme Vanz <jguilhermevanz@suse.com>

jvanz committed:
fix: upgrade-responder usage.

Updates the RFC to describe how the Longhord upgrade-responder will be
used to handle the Kubewarden metrics.

Signed-off-by: José Guilherme Vanz <jguilhermevanz@suse.com>

jvanz committed:
fix: user story and PR link

Updates the RFC 0023 adding the PR link and a required user story.

Signed-off-by: José Guilherme Vanz <jguilhermevanz@suse.com>

jvanz committed:
fix: specify telemetry is opt-out by default

Updates the RFC to state that stack telemetry will be enabled by default
and that instructions on how to disable it will be provided.

Signed-off-by: José Guilherme Vanz <jguilhermevanz@suse.com>

davideiori1 commented:
Hi.  I suggest adding a new metric to list what are the policies in use. This is to enable us to rank the most used policies vs the least used ones. This should be done only for the default policies provided by Kubewarden so that we do not track the names of the custom policies created by users.

viccuad commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,373 @@
+environments where Kubewarden is deployed. This data will help guide the
+project's future development.
+
+Kubearden stack telemetry will be enabled by default, but users will have an
```
```suggestion
Kubewarden stack telemetry will be enabled by default, but users will have an
```

viccuad commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,373 @@
+project's future development.
+
+Kubearden stack telemetry will be enabled by default, but users will have an
+easy way to disable it
```
```suggestion
easy way to disable it.
```

viccuad commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,373 @@
+data-driven decisions about the project's future. For this reason, we would
+like to add a feature to the Kubewarden stack that allows sending
+non-sensitive, anonymous environment information to a central service for
+aggregation and analysis
```
```suggestion
aggregation and analysis.
```

viccuad commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,373 @@
+As a Kubewarden user, I want to ensure that the process of sending telemetry
+data does not consume excessive resources in my environment.
+
+As a Kubewarden user, I want an easy way to disable the reconciler which send
```
I would reword and don't mention the reconciler nor a central location, just "an easy way to disable sending telemetry data".
The rest of user stories don't mention the reconciler, we haven't presented it yet, nor it is needed for understanding the story.

flavio replied:
I agree with Victor, the reconciler is a technical implementation, what the user cares is the end result: have an easy way to opt-out form telemetry

viccuad commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,373 @@
+
+This new periodic reconciler will start a time.Ticker that will periodically:
+
+- Collect information about how many PolicyServers are deployed
```
There's a mix of phrases ending in dots and missing them throughout the document, I would like to have consistency.

viccuad commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,373 @@
+- Build a JSON payload with the collected data and send it to the remote server
+  using secure communication.
+- The controller logs the available updates for the Kubewarden version running returned by
+  the `upgrade-responder` server
```
Here is the first mention of the upgrade-responder feature. I would like to mention in the user stories that the user would get a (configurable) notification of new Kubewarden releases.

Edit:
this is the first and only mention, as it isn't listed on the controller nor helm chart changes, nor explained how the user would receive notifications. Maybe is a left over?

Since we expect users to manage the Kubewarden stack via Gitops, maybe this feature is not of much value, and can be dropped off.

flavio replied:
Also, add a link to the `upgrade-responder` open source project, I know this is mentioned on the next section but it won't harm to have it here too (given it's the 1st time the reader jumps into that)

jvanz replied:
If we remove the warning from the logs, the reconciler becomes unnecessary. Its goal is to provide users with useful information when they share telemetry data. If the reconciler does nothing with the response about available upgrades, there is no reason to include it. It would just be sending data with no clear benefit to the user. In that case, users could simply query the upgrade-responder from another application without sending metrics in the process

viccuad replied:
I'm ok keeping the "update notification" feature. But then, I'm not sure if the controller logs is a nice medium to serve that notification.

viccuad commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,373 @@
+- `appVersion`: The Kubewarden version (e.g., v1.28.0).
+- `kubernetesVersion`: The Kubernetes version where the controller is running
+  (e.g., v1.33.0).
+- `instanceUid`: A UID to uniquely identify the controller instance.
```
I believe we need to explain how this `instanceUid` is unique. Is it unique per kubewarden-controller Deployment, or Kubewarden stack release (helm release)? 

I would expect it to be the second one. This allows us to not get duplicated data from several kubewarden-controller Deployments or restarts, and keeps consistent throughout upgrades.
Hence I believe it should be a random UID value created on `helm install`, not overwritten with `helm upgrade`, and saved as an opaque value in the Helm release ConfigMaps or Secrets.

We should also mention it in its relevant section (for example "Helm chart changes").

jvanz replied:
Yes, the idea was to use to identify kubewarden stack installation. 

I don't like the idea to have an opaque value created by helm. It's solve the issu. But, to me, it looks a too complicated solution for a simple uid generation.  I was reviewing the longhorn code and I have an idea. In the longhord request sending the telemetry data, it sends the longhorn [namespace uid](https://github.com/longhorn/longhorn/blob/9af0bd566dc222cf54055193d119305c0f79496e/dev/upgrade-responder/install.sh#L266). That's solve our issue here as well. It's a unique UID for each Kubewarden installation. And, we do not need to add another resource in our Helm chart templates.

viccuad replied:
That sounds good enough to me. I suspect Helm creates a UID per release too, as it needs to know when to upgrade them. But then we would need to configure our charts with RBAC to read the Helm Secrets or ConfigMaps. Your suggestion sounds better indeed.

viccuad commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,373 @@
+
+- `policyServerCount`: The number of PolicyServers deployed in the cluster.
+- `policiesCount`: The number of policies deployed in the cluster.
+- `value`: A field always set to `1`,
```
Could we explain why the field is always set to `1`? (looked at the upgrade-responder but it was not inmediately obvious, and either the reader should know why, or not care why).

viccuad commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,373 @@
+telemetry will be maintained by the SUSE team. All members of the Kubewarden
+organization on GitHub will have access to the collected data.
+
+## Helm Chart Changes
```
I would move this section upwards, after the "Controller changes" section. Hence one has the full changes in mind, to read how the infrastructure will interact with them.

viccuad commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,373 @@
+
+Alongside the `upgrade-responder` and InfluxDB, a Grafana instance with a custom
+dashboard will be configured. The dashboard will use InfluxDB as its data
+source and provide queries to visualize the desired metrics.
```
Do we intend to have this dashboard public, or only accessible to Kubewarden maintainers?
If the second one (the usual one), I believe it would be good to document this here, and how the login is managed, as this seems to be part of the CNCF legal review.

viccuad commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,373 @@
+
+The `upgrade-responder`, InfluxDB, and Grafana instances used for Kubewarden
+telemetry will be maintained by the SUSE team. All members of the Kubewarden
+organization on GitHub will have access to the collected data.
```
I would like to be specific here, maybe even linking to the [CONTRIBUTOR_LADDER.md](https://github.com/kubewarden/community/blob/main/CONTRIBUTOR_LADDER.md). Should it be only Kubewarden maintainers?

jvanz replied:
I'm not sure if this should be only for maintainers. The CNCF folks should have access to this information as well. In the end of the day, they are the "owners" of the project. Therefore, they should have access to this information.

viccuad replied:
Then, we should mention this here. Maybe "people with access to the cncf-kubewarden-maintainers at lists.cncf.io" is the set we want, as that includes both.

viccuad replied:
I still believe that the current wording is ambiguous. By "members of the Kubewarden org" do we mean those listed in https://github.com/orgs/kubewarden/people ? If so, I would link that there then.

Personally, I would be happier with "people with access to the cncf-kubewarden-maintainers at lists.cncf.io", which follows the approach of other projects we have reviewed. It may also help users in feeling that their data (even if it only contains metadata that is anonymous) is better guarded.

jvanz replied:
I've added your request in the document.

viccuad replied:
Many thanks! resolving.

viccuad commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,373 @@
+controller via the corresponding CLI flags.
+
+The `stackTelemetry.enabled` setting will be `true` by default. Users will be
+notified during the Kubewarden stack upgrade that this telemetry is included,
```
```suggestion
notified during the Kubewarden stack install and upgrade via Helm NOTES.txt that this telemetry is included,
```

flavio replied:
Good suggestion

viccuad reviewed (commented):
Thanks for the RFC!
I generally like it; I would like to clarify some points (the `instanceUID`) and access to processed data, plus I have some style and ordering quibbles though.

flavio commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,373 @@
+- Average number of policies per PolicyServer
+- Number of Kubewarden installations by version
+- Number of Kubewarden installations by Kubernetes version
+
```
I think it would be useful to have statistics about the policies being used. We should emit statistics only for the policies that have been authored by the Kubewarden project. We should not track policies created by the users, because that might expose some sensitive details.

We could do something simple, like sending the data only if the policy has been downloaded from the OCI registry of the Kubewarden project.

jvanz replied:
Good point. We can definitely implement this, but we need to be careful with the data model to avoid performance issues in InfluxDB over time.

The core challenge is tracking a dynamic list of official policies. Storing the names as a single concatenated string in a field ("policy1,policy2") or as individual fields with boolean values ("policy1": true) could make querying difficult and inefficient. Since fields aren't indexed, we'd need to do slow, full data scans to get any meaningful insights. Considering these approaches, these could be examples of the requests payload:


```json
{
  "appVersion": "v1.28.0",
  "extraTagInfo": {
    "kubernetesVersion": "v1.33.0",
  },
  "extraFieldInfo": {
    "policyServerCount": 2,
    "policiesCount": 3,
    "policiesInUse": "ghcr.io/kubewarden/policies/pod-privileged:latest,ghcr.io/kubewarden/policies/user-group-psp:v1.5.0",
    "namespaceUid": "<namespace-uid-where-kubewarden-is-installed>"
  }
```

Or

```json
{
  "appVersion": "v1.28.0",
  "extraTagInfo": {
    "kubernetesVersion": "v1.33.0",
  },
  "extraFieldInfo": {
    "policyServerCount": 2,
    "policiesCount": 3,
    "ghcr.io/kubewarden/policies/pod-privileged:latest": true,
    "ghcr.io/kubewarden/policies/user-group-psp:v1.5.0": true,
    "namespaceUid": "<namespace-uid-where-kubewarden-is-installed>"
  }
```

A better approach would be to add each policy name as a tag. This would allow for fast filtering and grouping. However, this could lead to a high series cardinality problem. Every unique combination of tags creates a new series. As users add or remove official policies, new series are created, leaving old ones dormant. Over time, this could lead to high memory usage and degraded performance. Again, some example:

```json
{
  "appVersion": "v1.28.0",
  "extraTagInfo": {
    "kubernetesVersion": "v1.33.0",
    "ghcr.io/kubewarden/policies/pod-privileged:latest": "true",
    "ghcr.io/kubewarden/policies/user-group-psp:v1.5.0": "true",
  },
  "extraFieldInfo": {
    "policyServerCount": 2,
    "policiesCount": 3,
    "namespaceUid": "<namespace-uid-where-kubewarden-is-installed>"
  }
```

One alternative approach would be to send in the metrics the number of official policies in use. Like this:

```json
{
  "appVersion": "v1.28.0",
  "extraTagInfo": {
    "kubernetesVersion": "v1.33.0",
  },
  "extraFieldInfo": {
    "policyServerCount": 2,
    "policiesCount": 3,
    "officialPoliciesCount": 2,
    "namespaceUid": "<namespace-uid-where-kubewarden-is-installed>"
  }

```

With this approach, we loose the ability to know the individual policies in use. But we can know, at least, the amount of them related to the number of total policies deployed.

jvanz replied:
Another ideia is to have a dedicated metric to keep track of  policy in use. 

```json
{
  "appVersion": "v1.28.0",
  "extraTagInfo": {
    "policyName": "ghcr.io/kubewarden/policies/pod-privileged:latest"
  },
  "extraFieldInfo": {  }
```

That way, we reduce the number of tag in the metrics, witch mitigate the  cardinality problem. And we do not loose the ability to group and filter the data. However, this required another instance of `upgrade-responder` and more request to be sent by the controller (1 for generic metrics + 1 for each policy)

jvanz replied:
If we want to rank the policies as sad by @davideiori1 , I believe we should go with the last option (dedicated metric) or the second option (the one of using policy names in tags but with the possible high cardinality problem)

davideiori1 replied:
@jvanz this is indeed a request from my side but I do not want to introduce any major refactoring to accomplish that. If that is the case, I suggest **we keep it simple** for now and record my request as backlog item for future enhancements.

flavio replied:
I think we don't need to argue about the low level format we will use to expose this information, this can be discussed at a later stage when the actual development will start.

What is important here is to agree on the data we plan to send and get an approval for that.

jvanz replied:
The thing is that data that we send to `upgrade-responder` is the data that will land in the InfluxDB. Therefore, it affect the queries that we will able to perform and by consequence the metrics we are able to track. That's why I've shared detailed options. 

This means that if we want to rank policies the controller probably should fetch/manipulate more information by the policies which affect the changes described in this document. If we do not mentioned that, we could hide  required actions.  Therefore, I see two options:

1. Do not mentions policies metrics at all and keep that for a second iteration of the metrics
2. Mention the suggested metric to here and add a text later in the  controller changes stating something like "fetch all the required data to allow the proposed metrics" 

I choose option 2.

flavio commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,373 @@
+- When new metadata or more metric data is required, it necessitates code
+  changes and a new release for both the controller and, maybe, the telemetry
+  server. The JSON schema versions must be kept in sync.
+- We are solely responsible for ensuring secure communication (TLS,
+  authentication, etc.) with the remote telemetry server.
```
Is that a drawback? I think this implicit with all telemetry solutions.

The TLS certificate used by the server is going to be issued by a trusted CA that is part of the standard `ca-certificates` bundles.

As for the authorization, this must be in place only for the grafana dashboard, but this will delegate authentication to GitHub

jvanz replied:
That's left over from the comparison with the Otel approach. Where we just need to do a configuration change instead of coding that in the controller.

flavio commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,373 @@
+
+The following subsection is a option given during the write of this RFC. But it
+was not select by the majority of the team. The team decided to move with the
+`upgrade-responder` option because the people that which will maintain the
+infrastructure more familiarized with the tool.
```
```suggestion
`upgrade-responder` option because this is a solution used by other
CNCF projects (Longhorn) and we (SUSE Rancher - who is willing to
pay for the infrastructure) also have experience managing and
maintaining the server-side infrastructure.
```

flavio committed:
Apply suggestions from code review

Co-authored-by: Víctor Cuadrado Juan <2196685+viccuad@users.noreply.github.com>
Signed-off-by: Flavio Castelli <flavio@castelli.me>

flavio reviewed (approved):
I like the proposal. I've left some minor comments

flavio commented:
@l-mb: can you please take a look at our plan and educate us about what we would need to do to integrate with [scarf](https://about.scarf.sh/)?

jvanz committed:
chore: addressing review comments

This commit does some changes requested during the review process

Signed-off-by: José Guilherme Vanz <jguilhermevanz@suse.com>

jvanz committed:
fix: namespace uid

Udpates the RFC adding a new field to store the namespace UID in the metric. Allowing the the identification of unique Kubewarden installation

Signed-off-by: José Guilherme Vanz <jguilhermevanz@suse.com>

jvanz committed:
fix: small text fixes

Small text fixes in the text

Signed-off-by: José Guilherme Vanz <jguilhermevanz@suse.com>

l-mb commented:
@flavio I have a pending PR for `longhorn/upgrade-responder` that'd allow us to then mirror incoming upgrade-responder events to said analytics platform as part of the ingestion pipeline, with no changes to how the upgrade-responder clients interact.

So having kubewarden adopt that deployment pattern makes a lot of sense and would fit neatly with the framework. (Scarf.sh is a CNCF-approved solution.)

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,413 @@
+environment in which it is running. The goal is not to collect information
+about user applications, but rather to learn more about the typical
+environments where Kubewarden is deployed. This data will help guide the
+project's future development.
```
```suggestion
This RFC describes how the Kubewarden stack will send telemetry data about the
environment in which it is running. The goal isn't to collect information
about user applications, but rather to learn more about typical Kubewarden deployment
environments. This data will help guide the
project's future development.
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,413 @@
+# Detailed design
+
+> [!WARNING]
+> As a CNCF project, it is necessary to get approval from the CNCF to implement either of the following options.
```
```suggestion
> As a CNCF project, it's necessary to get approval from the CNCF to implement either of the following options.
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,413 @@
+
+The main changes to the Kubewarden controller would be:
+
+- A configuration option to specify the endpoint that will receive the
```
```suggestion
- A configuration option to specify the endpoint that receives the
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,413 @@
+### Controller Configuration
+
+The controller should have two new CLI flags to enable telemetry collection and
+delivery: `--stack-telemetry-endpoint` and `--stack-telemetry-period`. The first
```
```suggestion
delivery: `--stack-telemetry-endpoint` and `--stack-telemetry-period`. The former
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
````diff
@@ -0,0 +1,413 @@
+manager --leader-elect ... --stack-telemetry-endpoint="https://metrics.kubewarden.io/" --stack-telemetry-period=6h ...
+```
+
+When the `--stack-telemetry-endpoint` flag is not defined, no telemetry will be
````
```suggestion
When the `--stack-telemetry-endpoint` flag isn't defined, no telemetry will be
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,413 @@
+
+### Controller Information Collection and Delivery
+
+When the telemetry endpoint is configured, a new periodic job should be created
```
```suggestion
When configuring the telemetry endpoint, a new periodic job should be created
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,413 @@
+using the Runnable interface from the controller-runtime package and integrated
+into the manager alongside the existing reconcilers.
+
+This new periodic reconciler will start a time.Ticker that will periodically:
```
```suggestion
This new periodic reconciler will start a `time.Ticker` that will periodically:
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,413 @@
+
+The telemetry server will be an instance of the [Longhorn
+upgrade-responder](https://github.com/longhorn/upgrade-responder).
+This server will receive information from controllers and store it in an
```
```suggestion
This server receives information from controllers and store it in an
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,413 @@
+- `policyServerCount`: The number of PolicyServers deployed in the cluster.
+- `policiesCount`: The number of policies deployed in the cluster.
+- `namespaceUid`: A UID to uniquely identify the Kubewarden stack installation.
+  This should be UID of the namespace where Kubewarden is installed.
```
```suggestion
  This should be UID of the Kubewarden installation namespace.
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,413 @@
+
+  ### Request validation schema
+
+  upgrade-responder should be configured with a JSON schema used to validate the
```
Not sure about this one but how about:
```suggestion
  The `upgrade-responder` should be configured with a JSON schema used to validate the
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,413 @@
+
+- When new metadata or more metric data is required, it necessitates code
+  changes and a new release for both the controller and, maybe, the telemetry
+  server. The JSON schema versions must be kept in sync.
```
```suggestion
- When requiring new metadata or more metric data, code
  changes will be needed. Along with a new release for both the controller and, possibly, the telemetry
  server. The JSON schema versions must be kept synchronized.
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,413 @@
+`upgrade-responder` option because this is a solution used by other CNCF
+projects (Longhorn) and we (SUSE Rancher - who is willing to pay for the
+infrastructure) also have experience managing and maintaining the server-side
+infrastructure
```
```suggestion
The following subsection is an option provided during the writing of this RFC. But, it
was not selected by a majority of the team. The team decided to move forward with the
`upgrade-responder` option because this is a solution used by other CNCF
projects (Longhorn). Also, we (SUSE Rancher — who is willing to pay for the
infrastructure) have experience managing and maintaining the server-side
infrastructure.
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,413 @@
+## Not selected option: OpenTelemetry Integration
+
+This implementation option aims for a less intrusive approach, decoupling the
+controller from the telemetry backend. With this design, all telemetry will be
```
```suggestion
controller from the telemetry back-end. With this design, all telemetry will be
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,413 @@
+controller from the telemetry backend. With this design, all telemetry will be
+managed outside the Kubewarden controller's reconciliation loop. The controller
+will be instrumented with telemetry collection points that send data to an
+OpenTelemetry collector, which will then be configured to forward this data to
```
```suggestion
OpenTelemetry collector, which will be configured to forward this data to
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
````diff
@@ -0,0 +1,413 @@
+policiesGauge.Record(context.Background(), 1, metric.WithAttributes(commonAttrs...)) // Corrected context.Background()
+```
+
+Once the metric is recorded, it is up to the OpenTelemetry collector to
````
```suggestion
Once the metric is recorded, it's up to the OpenTelemetry collector to
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,413 @@
+
+### OpenTelemetry Configuration
+
+In the OpenTelemetry collector configuration, a new pipeline should be added to
```
```suggestion
In the OpenTelemetry collector configuration, a new pipeline will be added to
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,413 @@
+[processors](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor),
+potentially with no changes to the controller code. This makes telemetry
+customization more flexible. In the example above, processors are used to add
+labels, filter unrelated metrics, and batch metrics before sending.
```
```suggestion
This OpenTelemetry configuration can be used to control the telemetry export
and, be expanded over time to include more metadata by leveraging the rich
ecosystem of OpenTelemetry
[processors](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor).
This would potentially need no changes to the controller code. This makes telemetry
customization more flexible. In the example above, processors will be used to add
labels, filter unrelated metrics, and batch metrics before sending.
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,413 @@
+All OpenTelemetry collector configuration can be managed in the Helm chart
+installation/update. The remote endpoint would be provided in the Helm Chart
+values. If the `stackTelemetry.enabled` is disabled, the OpenTelemetry
+configuration will not be applied:
```
```suggestion
configuration won't be applied:
```

jhkrug commented on rfc/0023-kubewarden-stack-telemetry.md:
```diff
@@ -0,0 +1,413 @@
+Processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/geoipprocessor)
+to add location information to the received metrics.
+
+# Unresolved Questions
```
```suggestion
# Unresolved questions
```

jhkrug reviewed (changes requested):
Some small suggestions or improvements to consider. Thanks.

viccuad reviewed (commented):
The RFC is looking good to me; only missing a clarification on exactly who gets to access the telemetry metadata, for my approval.

jvanz committed:
Apply suggestions from code review

Co-authored-by: John Krug <jh.krug@gmail.com>
Signed-off-by: José Guilherme Vanz <jguilhermevanz@suse.com>

flavio reviewed (approved):
The RFC looks good to me, I'm fine merging it and submitting to CNCF for their approval

viccuad reviewed (approved):
LGTM too!

robertsirc reviewed (approved):
LGTM, I like the direction of this.
